rfcs/0164-worker-machine-image-building-automation.md

```diff
@@ -0,0 +1,203 @@
+  create tasks for this worker type
+* (Optional) we add a policy to taskcluster-auth that enforces that the
+  following scopes:
+
+    * `queue:create-task:highest:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:very-high:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:high:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:medium:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:low:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:very-low:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:lowest:proj-taskcluster/worker-image-builder`
+    * `queue:create-task:normal:proj-taskcluster/worker-image-builder`
+
+  can only be present in the expanded scopes of the following roles and clients:
+
+    * role `repo:github.com/mozilla/community-tc-config:branch:*`
+    * client `static/taskcluster/github`
+
+  Note, client `static/taskcluster/queue` shouldn't require such scopes, unless
+  we are forced to create the `proj-taskcluster.worker-image-builder` tasks from
+  a decision task. The author is currently unaware of a reason this would be
+  required.
+* The worker type is *manually* secured with a chain of trust ed25519 key by a
+  trusted employee
+* Access to image building workers is restricted to only a small whitelist of
+  trusted employees
+* The public chain of trust key for the worker image builder is published to
+  one or more public places
+* The community-tc-config `.taskcluster.yml` file will trigger a unique task
+  for each image set in the `imagesets` directory of the community-tc-config
+  repo, for any __PR commits__ to validate the change. This task may be a noop
+  (see later).
+* The community-tc-config `.taskcluster.yml` file will trigger a different task
+  for any __PR merge commits__ made against `main` branch (more about this
+  later).
+* The script which the worker-image-builder script-worker worker runs for PR
+  commits:
+  * Accepts a payload that only contains an image set name, and the git commit
+    SHA of the PR commit
+  * Executes a script which is hard-wired into the worker, which determines
+    whether the commit is a valid PR commit that is a decescendant of
+    the commit from which the latest version of the image set was built from, and
+    whether, there are any possible changes to the image set build since the most
+    recent build for that image set in the current PR, or for the image set if
+    there aren't already images built for the current PR (explained below).
+  * For this, a reliable mechanism will be needed that determines which commit
+    of community-tc-config a given imageset is built from. Perhaps this mapping
+    from community-tc-config git commit SHA to imageset names is persisted in
+    community-tc-config repo directly.
+  * If there are changes to any files which are used to produce the image set
+    images, a new image set will be built, with name `<imageset>-<PR>`. This
```

Comment:
For PRs with multiple revisions, how can we be sure that a given worker running `<imageset>-<PR>` has the most recent changes from that PR? Is there a piece of metadata that will contain the hash of the most recent commit contained in the image?